nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java

```diff
@@ -0,0 +1,276 @@
+        coll.createIndex(new Document(TTL_KEY, ONE), indexOptions);
+    }
+
+    @Override
+    public void close() {
+        if (coll != null) {
+            coll = null;
+        }
+    }
+
+    @Override
+    public void clear() {
+        if (log.isTraceEnabled()) {
+            log.trace("MongoDB: CLEAR");
+        }
+        coll.deleteMany(new Document());
+    }
+
+    // if possible, store the bytes as a UTF-8 string
+    protected static Object toStorage(byte[] bytes) {
+        try {
+            return UTF_8_DECODERS.get().decode(ByteBuffer.wrap(bytes)).toString();
+        } catch (CharacterCodingException e) {
+            // could not decode as UTF-8, use a binary
+            return new Binary(bytes);
+        }
+    }
+
+    protected static byte[] fromStorage(Object value) {
+        if (value instanceof Binary) {
+            return ((Binary) value).getData();
+        } else if (value instanceof String) {
+            return ((String) value).getBytes(UTF_8);
+        } else {
+            throw new UnsupportedOperationException(value.getClass().getName());
+        }
+    }
+
+    @Override
+    public byte[] get(String key) {
+        Document doc = coll.find(Filters.eq(ID_KEY, key)).first();
+        if (doc == null) {
+            if (log.isTraceEnabled()) {
+                log.trace("MongoDB: GET " + key + " = null");
+            }
+            return null;
+        }
+        Object value = doc.get(VALUE_KEY);
+        if (log.isTraceEnabled()) {
+            log.trace("MongoDB: GET " + key + " = " + value);
+        }
+        return fromStorage(value);
+    }
+
+    @Override
+    public void put(String key, byte[] bytes, long ttl) {
+        if (bytes == null) {
+            if (log.isTraceEnabled()) {
+                log.trace("MongoDB: DEL " + key);
+            }
+            coll.deleteOne(new Document(ID_KEY, key));
+        } else {
+            Document filter = new Document(ID_KEY, key);
+            Object value = toStorage(bytes);
+            Document doc = new Document(VALUE_KEY, value);
+            setTTL(doc, ttl);
+            if (log.isTraceEnabled()) {
+                log.trace("MongoDB: PUT " + key + " = " + value + (ttl == 0 ? "" : " (TTL " + ttl + ")"));
+            }
+            coll.replaceOne(filter, doc, new UpdateOptions().upsert(true));
+        }
+    }
+
+    protected void setTTL(Document doc, long ttl) {
+        if (ttl != 0) {
+            doc.append(TTL_KEY, new Date(System.currentTimeMillis() + ttl * 1000));
+        }
+    }
+
+    @Override
+    public boolean setTTL(String key, long ttl) {
+        Document filter = new Document(ID_KEY, key);
+        Document update;
+        if (ttl == 0) {
+            update = new Document(MONGODB_UNSET, new Document(TTL_KEY, ONE));
+        } else {
+            Document doc = new Document();
+            setTTL(doc, ttl);
+            update = new Document(MONGODB_SET, doc);
```

Comment:
Could be replaced with `update = com.mongodb.client.model.Updates.set(TTL_KEY, ONE)`.